qiskit/synthesis/mcx_synthesis.py

```diff
@@ -0,0 +1,273 @@
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""MCX synthesis."""
+
+from abc import ABC, abstractmethod
+from math import ceil
+import numpy
+
+from qiskit.circuit.quantumregister import QuantumRegister
+from qiskit.circuit.library import C3XGate, C4XGate, U2Gate, RCCXGate, U1Gate, CXGate, CCXGate, HGate, MCU1Gate, MCXGate
+
+
+class MCXSynthesis(ABC):
+    """Interface for MCX synthesis algorithms."""
+
+    @abstractmethod
+    def synthesize(self, num_cntl_qubits):
+        """Synthesize..
+
+        Args:
+
+        Returns:
+            QuantumCircuit: A circuit implementing the evolution.
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    @abstractmethod
+    def get_num_ancilla_qubits(num_ctrl_qubits):
+        """Returns the number of ancilla qubits required by the synthesis algorithm."""
+
+        raise NotImplementedError
+
+
+class MCXSynthesisGrayCode(MCXSynthesis):
+    r"""Implement the multi-controlled X gate using the Gray code.
+
+    This delegates the implementation to the MCU1 gate, since :math:`X = H \cdot U1(\pi) \cdot H`.
+    """
+
+    def __init__(self):
+        self.name = "noancilla"
+
+    @staticmethod
+    def get_num_ancilla_qubits(num_ctrl_qubits):
+        return 0
+
+    def synthesize(self, num_ctrl_qubits):
+        """Define the MCX gate using the Gray code."""
+        # pylint: disable=cyclic-import
+        from qiskit.circuit.quantumcircuit import QuantumCircuit
+
+        num_qubits = num_ctrl_qubits + self.get_num_ancilla_qubits(num_ctrl_qubits) + 1
+        q = QuantumRegister(num_qubits, name="q")
+        qc = QuantumCircuit(q, name=self.name)
+        qc._append(HGate(), [q[-1]], [])
+        qc._append(MCU1Gate(numpy.pi, num_ctrl_qubits=num_ctrl_qubits), q[:], [])
+        qc._append(HGate(), [q[-1]], [])
+
+        return qc
+
+
+class MCXSynthesisRecursive(MCXSynthesis):
+    """Implement the multi-controlled X gate using recursion.
+
+    Using a single ancilla qubit, the multi-controlled X gate is recursively split onto
+    four sub-registers. This is done until we reach the 3- or 4-controlled X gate since
+    for these we have a concrete implementation that do not require ancillas.
+    """
+
+    def __init__(self):
+        self.name = "mcx_recursive"
+
+    @staticmethod
+    def get_num_ancilla_qubits(num_ctrl_qubits):
+        return int(num_ctrl_qubits > 4)
+
+    def synthesize(self, num_ctrl_qubits):
+        """Define the MCX gate using recursion."""
+
+        # pylint: disable=cyclic-import
+        from qiskit.circuit.quantumcircuit import QuantumCircuit
+        num_qubits = num_ctrl_qubits + self.get_num_ancilla_qubits(num_ctrl_qubits) + 1
```

Comment:
Could we use actual `AncillaRegister` for the ancilla qubits? That'll give us the `num_ancillas` property for free and provides more information to the compiler (which we cannot use yet but hopefully can in the future 🙂).

The same comment holds for the other synthesis algorithms, too.

Reply:
I don't fully understand this comment. Currently, `self.get_num_ancilla_qubits` returns the number of ancilla qubits required for implementing a given synthesis algorithm with a given number of control qubits. What do you mean by "actual AncillaRegisters"?

Reply:
I mean that we have a qubit type for ancillas, `AncillaQubit` which we should use here 🙂 For example
```python
>>> from qiskit.circuit import QuantumRegister, AncillaRegister, QuantumCircuit
>>> qr = QuantumRegister(3)
>>> ar = AncillaRegister(2)
>>> circuit = QuantumCircuit(qr, ar)
>>> circuit.draw()

q1_0:

q1_1:

q1_2:

a1_0:

a1_1:

>>> circuit.num_qubits
5
>>> circuit.num_ancillas
2
```
This way we expose the information that our circuit contains ancilla qubits. And in the future the compiler should be able to use this information and e.g. re-use ancilla qubits instead of allocating new ones.

Reply:
Thanks! I understand your point much better now, though I am still lost in the details. Currently, a (MCX) gate does not know about the circuit it belongs to, and no MCX synthesis algorithm takes the overall circuit into account. The `QuantumCircuit.mcx` function expects a sufficient number of ancilla qubits to be provided, and complains if the required number of ancilla qubits (computed based on the MCX mode string and the number of control qubits) is larger than that. Looking at the grover algorithm as a code example that creates MCX gates: it gets an MCX mode as a parameter, then calls a (static) function to find out how many ancilla qubits would be required (for this MCX mode and this number of control qubits), then would create the required ancilla qubits, and then finally call `mcx`. So I am not quite sure what to do here. Ideally, we also probably do not need to specify the MCX synthesis algorithm from the start, and should be able to choose the best-suited algorithm only when we need to synthesize the gate (with the best synthesis algorithm possibly depending on the number of ancilla qubits actually available).

Reply:
Ah yes, it's a staticmethod. I got confused because in this context it was called as `self.get_num_ancilla_qubits`! 

Then my comment simplifies a bit, we need to keep that staticmethod but I think it would still be good to use an `AncillaRegister` for the ancilla qubits within the definition. Concretely I mean:
```python
qr = QuantumRegister(num_ctrl_qubits + 1, name="q")
ar = AncillaRegister(self.get_num_ancilla_qubits(num_ctrl_qubits), name="a")
qc = QuantumCircuit(qr, ar, name=self.name)
```